src/network/cni/iptables.rs

```diff
@@ -0,0 +1,1177 @@
+//! Linux iptables interface
+
+use crate::network::cni::port::PortMapping;
+use anyhow::{bail, Context, Result};
+use async_trait::async_trait;
+use derive_builder::Builder;
+use dyn_clone::{clone_trait_object, DynClone};
+use getset::{CopyGetters, Getters, Setters};
+use ipnetwork::IpNetwork;
+use log::trace;
+use std::{
+    collections::BTreeMap,
+    fmt::{self, Debug},
+    net::Ipv4Addr,
+    path::PathBuf,
+    process::Output,
+};
+use tokio::process::Command;
+
+#[async_trait]
+/// Iptables behavior trait.
+pub trait Iptables: DynClone + Send + Sync {
+    /// Setup the provided `Chain` with respect of whether to use IPv6.
+    async fn setup(&self, _chain: &Chain) -> Result<()> {
+        Ok(())
+    }
+
+    /// Idempotently delete a chain. It will not error if the chain doesn't exist. It will first
+    /// delete all references to this chain in the `entry_chains`.
+    async fn teardown(&self, _chain: &Chain) -> Result<()> {
+        Ok(())
+    }
+
+    /// Enable IPv6 usage for this iptables instance.
+    fn set_ipv6(&mut self, _value: bool) {}
+}
+
+clone_trait_object!(Iptables);
+
+#[derive(Builder, Clone, CopyGetters, Getters, Setters)]
+#[builder(pattern = "owned", setter(into))]
+/// The main interface to the Linux iptables.
+pub struct DefaultIptables {
+    #[get]
+    /// Path to the `iptables` binary.
+    iptables_binary: PathBuf,
+
+    #[get]
+    /// Path to the `ip6tables` binary.
+    ip6tables_binary: PathBuf,
+
+    #[getset(set, get_copy)]
+    #[builder(default = "false")]
+    /// Whether to use IPv6 or not.
+    use_ipv6: bool,
+
+    #[getset(set, get)]
+    #[builder(private, default = "Box::new(DefaultExec)")]
+    /// Internal command executor to be used
+    exec: Box<dyn Exec>,
+}
+
+#[async_trait]
+impl Iptables for DefaultIptables {
+    async fn setup(&self, chain: &Chain) -> Result<()> {
+        self.chain_setup(chain).await
+    }
+
+    async fn teardown(&self, chain: &Chain) -> Result<()> {
+        self.chain_teardown(chain).await
+    }
+
+    fn set_ipv6(&mut self, value: bool) {
```

Comment:
Is this something that should be bool as part of `setup`?

Reply:
I thought about it, and we would also need it for `teardown`. Then we would have to pass around the value down to `run`, which seems a bit overkill to me. My main reason to add this method was to not hat a need to lookup the binary at every invocation.